lib/TPP/Dialect/Tpp/BufferizableOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,372 @@
+//===- BufferizableOpInterfaceImpl.cpp - Impl. of BufferizableOpInterface -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Tpp/TppDialect.h"
+#include "TPP/Dialect/Tpp/TppOps.h"
+#include "mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Operation.h"
+
+using namespace mlir;
+using namespace mlir::bufferization;
+
+namespace mlir {
+namespace tpp {
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Unary
+//===----------------------------------------------------------------------===//
+
+// Helper function to bufferize unary operations.
+template <typename OpTy>
+static LogicalResult bufferizeUnaryOp(Operation *op, RewriterBase &rewriter,
+                                      const BufferizationOptions &options) {
+  auto unaryOp = cast<OpTy>(op);
+  auto loc = unaryOp.getLoc();
+  FailureOr<Value> buffer =
+      getBuffer(rewriter, unaryOp.getInputs()[0], options);
+  if (failed(buffer))
+    return failure();
+  // Out-of-place bufferization.
+  if (unaryOp.getInputs()[0].getType() != unaryOp.getResultType()) {
+    FailureOr<Value> alloc =
+        allocateTensorForShapedValue(rewriter, loc, unaryOp.getResult(0),
+                                     /*escape=*/true, options, /*copy=*/false);
+    if (failed(alloc))
+      return failure();
+    FailureOr<Value> allocBuffer = getBuffer(rewriter, *alloc, options);
+    if (failed(allocBuffer))
+      return failure();
+    rewriter.create<OpTy>(loc, *buffer, *allocBuffer);
+    replaceOpWithBufferizedValues(rewriter, op, *allocBuffer);
+    return success();
+  }
+  // In-place bufferization.
+  rewriter.create<OpTy>(loc, *buffer, *buffer);
+  replaceOpWithBufferizedValues(rewriter, op, *buffer);
```

Comment:
In case the input tensor has users after the currently bufferized op, are there any safeguards to prevent in-place bufferization?
For example:
```
%0 = unary %A
%1 = matmul %0, %val
%2 = matmul  %A, %val  // unary needs out-of-place bufferization due to data dependency
```

Reply:
Bufferization will insert a copy

Reply:
I think I have an example somewhere in the test, will look for it

Reply:
You're right, I see some cases  with `add` where copy is inserted 👍